edgedb/protocol/codecs/codecs.pyx

```diff
@@ -798,6 +796,264 @@ cdef pgvector_decode(pgproto.CodecContext settings, FRBuffer *buf):
     return val
 
 
+# Needed to perform bit manipulations to convert from float16 to float32
+cdef union IntFloatSwap:
+    uint32_t i
+    float f
+
+
+# Optimization to convert f16 to f32
+cdef:
+    uint32_t[2048] MANTISSA_TABLE
+    uint32_t[64] EXP_TABLE
+    uint32_t[64] OFFSET_TABLE
+    uint32_t[512] BASE_TABLE
+    uint32_t[512] SHIFT_TABLE
+
+
+cdef convertmantissa(uint32_t n):
+    cdef:
+        uint32_t m
+        uint32_t e
+
+    m = n << 13                 # Zero pad mantissa bits
+    e = 0                       # Zero exponent
+    # While not normalized
+    while (m & 0x00800000) == 0:
+        e -= 0x00800000         # Decrement exponent (1<<23)
+        m <<= 1                 # Shift mantissa
+
+    m &= ~0x00800000            # Clear leading 1 bit
+    e += 0x38800000             # Adjust bias ((127-14)<<23)
+
+    return m | e
+
+
+cdef generate_tables():
```

Comment:
Could you put a comment in here (and the final commit message) with a cite to the algorithm you've implemented?

Reply:
Comment added